server/apps/immich/src/api-v1/oauth/oauth.service.spec.ts

```diff
@@ -0,0 +1,169 @@
+    jest.spyOn(generators, 'state').mockReturnValue('state');
+    jest.spyOn(Issuer, 'discover').mockResolvedValue({
+      id_token_signing_alg_values_supported: ['HS256'],
+      Client: jest.fn().mockResolvedValue({
+        issuer: {
+          metadata: {
+            end_session_endpoint: 'http://end-session-endpoint',
+          },
+        },
+        authorizationUrl: jest.fn().mockReturnValue('http://authorization-url'),
+        callbackParams: jest.fn().mockReturnValue({ state: 'state' }),
+        callback: jest.fn().mockReturnValue({ access_token: 'access-token' }),
+        userinfo: jest.fn().mockResolvedValue({ email }),
+      }),
+    } as any);
+
+    userRepositoryMock = {
+      get: jest.fn(),
+      getAdmin: jest.fn(),
+      getByEmail: jest.fn(),
+      getList: jest.fn(),
+      create: jest.fn(),
+      update: jest.fn(),
+      delete: jest.fn(),
+      restore: jest.fn(),
+    };
+
+    immichJwtServiceMock = {
+      getCookieNames: jest.fn(),
+      getCookies: jest.fn(),
+      createLoginResponse: jest.fn(),
+      validateToken: jest.fn(),
+      extractJwtFromHeader: jest.fn(),
+      extractJwtFromCookie: jest.fn(),
+    } as unknown as jest.Mocked<ImmichJwtService>;
+
+    configServiceMock = {
+      get: jest.fn(),
+    } as unknown as jest.Mocked<ConfigService>;
+
+    sut = new OAuthService(immichJwtServiceMock, configServiceMock, userRepositoryMock);
+  });
+
+  it('should be defined', () => {
+    expect(sut).toBeDefined();
+  });
+
+  describe('generateConfig', () => {
+    it('should work when oauth is not configured', async () => {
+      await expect(sut.generateConfig({ redirectUri: 'http://callback' })).resolves.toEqual({ enabled: false });
+      expect(configServiceMock.get).toHaveBeenCalled();
+    });
+
+    it('should generate the config', async () => {
+      configServiceMock.get.mockImplementation(
+        mockConfig({
+          OAUTH_ENABLED: true,
+          OAUTH_BUTTON_TEXT: 'OAuth',
+        }),
+      );
+      sut = new OAuthService(immichJwtServiceMock, configServiceMock, userRepositoryMock);
+      await expect(sut.generateConfig({ redirectUri: 'http://redirect' })).resolves.toEqual({
+        enabled: true,
+        buttonText: 'OAuth',
+        url: 'http://authorization-url',
+      });
+    });
+  });
+
+  describe('callback', () => {
+    it('should throw an error if OAuth is not enabled', async () => {
+      await expect(sut.callback({ url: '' })).rejects.toBeInstanceOf(BadRequestException);
+    });
+
+    it('should not allow auto registering', async () => {
+      configServiceMock.get.mockImplementation(
+        mockConfig({
+          OAUTH_ENABLED: true,
+          OAUTH_AUTO_REGISTER: false,
+        }),
+      );
+      sut = new OAuthService(immichJwtServiceMock, configServiceMock, userRepositoryMock);
+      jest.spyOn(sut['logger'], 'debug').mockImplementation(() => null);
+      jest.spyOn(sut['logger'], 'warn').mockImplementation(() => null);
```

Comment:
🤔 we shouldn't mock / spy `sut` methods. Instead we inject / mock dependencies. In this case we probably want to mock the `Logger` dependency

Reply:
Yeah, I'm not sure what to do here. I like using the logger this way because it adds a prefix to each log, but I agree injecting it would be better for testing. 

I could either use the static methods or try to figure out how to make it inject it in the constructor. 

Maybe I'll just use the static methods for now.

Reply:
Yeah, static method is probably the easiest and also easy to mock (apart from using dependency injection). You should also be able to mock the `Logger` constructor: [docs](https://jestjs.io/docs/es6-class-mocks)